### PR TITLE
s3 client

### PR DESCRIPTION
**Pull request by kyleroche**
fixed PR

### CODE REVIEW

File: griptape/tools/aws_s3_client/tool.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+from schema import Schema, Literal
+from attr import define, field, Factory
+from griptape.artifacts import BaseArtifact, TextArtifact, ErrorArtifact
+from griptape.core.decorators import activity
+from griptape.tools import BaseAwsClient
+
+@define
+class AwsS3Client(BaseAwsClient):
+    session: boto3.session = field(kw_only=True)
+    s3_client: boto3.client = field(
```
Comment (vasinov) [inline griptape/tools/aws_s3_client/tool.py:12]: Can you add `kw_only=True` to this field?

File: griptape/tools/aws_s3_client/tool.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        default=Factory(lambda self: self.session.client("s3"), takes_self=True)
+    )
+
+    @activity(config={
+        "description": "Retruns the access control list(ACL) of an bucket.",
```
Comment (vasinov) [inline griptape/tools/aws_s3_client/tool.py:17]: Change to "Can be used to get an access control list (ACL) of an AWS S3 bucket."

Remember, the LLM has no context about this tool outside of the activity description, so the description should always be comprehensive.

File: griptape/tools/aws_s3_client/tool.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        "schema": Schema({
+            Literal(
+                "Bucket",
```
Comment (vasinov) [inline griptape/tools/aws_s3_client/tool.py:20]: 1. Should be lower case.
2. Rename to `bucket_name`.

File: griptape/tools/aws_s3_client/tool.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+                description="The bucket name that contains the object for which to get the ACL information."
+            ): str
+        })
+    })
+    def get_bucket_acl(self, params: dict) -> TextArtifact | ErrorArtifact:
+        try:
+            acl = self.s3_client.get_bucket_acl(
+                Bucket=params["values"]["Bucket"]
```
Comment (vasinov) [inline griptape/tools/aws_s3_client/tool.py:28]: Rename to `bucket_name`.

File: griptape/tools/aws_s3_client/tool.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    s3_client: boto3.client = field(
+        default=Factory(lambda self: self.session.client("s3"), takes_self=True)
+    )
+
+    @activity(config={
+        "description": "Retruns the access control list(ACL) of an bucket.",
+        "schema": Schema({
+            Literal(
+                "Bucket",
+                description="The bucket name that contains the object for which to get the ACL information."
```
Comment (vasinov) [inline griptape/tools/aws_s3_client/tool.py:21]: Make this sound simpler for the LLM. Something like "Name of the AWS S3 bucket for which to get an ACL."

File: griptape/tools/aws_s3_client/tool.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+            )
+            return TextArtifact(acl)
+        except Exception as e:
+            return ErrorArtifact(f"error getting bucket acl: {e}")
+
+    @activity(config={
+        "description": "Returns the policy of a specified bucket",
+        "schema": Schema({
+            Literal(
+                "Bucket",
```
Comment (vasinov) [inline griptape/tools/aws_s3_client/tool.py:38]: Rename to `bucket_name`.

File: griptape/tools/aws_s3_client/tool.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+                description="The bucket name for which to get the bucket policy."
+            ): str
+        })
+    })
+    def get_bucket_policy(self, params: dict) -> TextArtifact | ErrorArtifact:
+        try:
+            policy = self.s3_client.get_bucket_policy(
+                Bucket=params["values"]["Bucket"]
```
Comment (vasinov) [inline griptape/tools/aws_s3_client/tool.py:46]: Rename to `bucket_name`.

File: griptape/tools/aws_s3_client/tool.py [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+            )
+            return TextArtifact(policy)
+        except Exception as e:
+            return ErrorArtifact(f"error getting bucket policy: {e}")
+
+    @activity(config={
+        "description":"Retruns the access control list(ACL) of an object.",
```
Comment (vasinov) [inline griptape/tools/aws_s3_client/tool.py:53]: Change to "Can be used to get an access control list (ACL) of an object in the AWS S3 bucket."

File: griptape/tools/aws_s3_client/tool.py [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+        "schema": Schema({
+            Literal(
+                "Object",
```
Comment (vasinov) [inline griptape/tools/aws_s3_client/tool.py:56]: Rename to `bucket_name`

File: griptape/tools/aws_s3_client/tool.py [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+        try:
+            acl = self.s3_client.get_bucket_acl(
+                Bucket=params["values"]["Bucket"]
+            )
+            return TextArtifact(acl)
+        except Exception as e:
+            return ErrorArtifact(f"error getting bucket acl: {e}")
+
+    @activity(config={
+        "description": "Returns the policy of a specified bucket",
```
Comment (vasinov) [inline griptape/tools/aws_s3_client/tool.py:35]: Change to "Can be used to get an AWS S3 bucket policy."

File: griptape/tools/aws_s3_client/tool.py [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+
+    @activity(config={
+        "description":"Retruns the access control list(ACL) of an object.",
+        "schema": Schema({
+            Literal(
+                "Object",
+                description="The bucket name that contains the object for which to get the ACL information."
+            ): str,
+            Literal(
+                "Key",
```
Comment (vasinov) [inline griptape/tools/aws_s3_client/tool.py:60]: Change to `object_key`.

File: griptape/tools/aws_s3_client/tool.py [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+            acl = self.s3_client.get_object_acl(
+                Bucket=params["values"]["Bucket"],
+                Key=params["values"]["Key"]
+            )
+            return TextArtifact(acl)
+        except Exception as e:
+            return ErrorArtifact(f"error getting object acl: {e}")
+
+    @activity(config={
+        "description": "Lists all aws s3 buckets."
```
Comment (vasinov) [inline griptape/tools/aws_s3_client/tool.py:76]: Change to "Can be used to list all AWS S3 buckets."

File: griptape/tools/aws_s3_client/tool.py [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+    })
+    def list_s3_buckets(self, params: dict) -> list[TextArtifact] | ErrorArtifact:
+        try:
+            buckets = self.s3_client.list_buckets()
```
Comment (vasinov) [inline griptape/tools/aws_s3_client/tool.py:80]: Does this automatically handle pagination?

File: griptape/tools/aws_s3_client/tool.py [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+            return [TextArtifact(str(b)) for b in buckets["Buckets"]]
+        except Exception as e:
+            return ErrorArtifact(f"error listing s3 buckets: {e}")
+
+    @activity(config={
+        "description": "Lists all objects in an S3 bucket.",
```
Comment (vasinov) [inline griptape/tools/aws_s3_client/tool.py:86]: Change to "Can be used to list all objects in an AWS S3 bucket."

File: griptape/tools/aws_s3_client/tool.py [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+        "schema": Schema({
+            Literal(
+                "Bucket",
+                description="The name of the S3 bucket to list."
+            ): str
+        })
+    })
+    def list_objects(self, params: dict) -> list[TextArtifact] | ErrorArtifact:
+        try:
+            objects = self.s3_client.list_objects_v2(
```
Comment (vasinov) [inline griptape/tools/aws_s3_client/tool.py:96]: Does this handle pagination?

File: griptape/tools/base_aws_client.py [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+from griptape.core import BaseTool
+from griptape.core.decorators import activity
+
+
+@define
+class BaseAwsClient(BaseTool, ABC):
+    session: boto3.session = field(kw_only=True)
+
+    @activity(config={
+        "description": "Returns the account and IAM principal of the AWS credentials being used."
```
Comment (vasinov) [inline griptape/tools/base_aws_client.py:14]: Change to "Can be used to get current AWS account and IAM principal."

**Review by vasinov — CHANGES_REQUESTED** [review]

File: griptape/tools/aws_s3_client/tool.py (hunk 13 above)
Comment (kyleroche) [reply griptape/tools/aws_s3_client/tool.py:80]: https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3/client/list_objects_v2.html

need to check IsTruncated and add another method for retrieval

File: griptape/tools/aws_s3_client/tool.py (hunk 15 above)
Comment (kyleroche) [reply griptape/tools/aws_s3_client/tool.py:96]: https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3/client/list_objects_v2.html

need to check IsTruncated and add another method for retrieval

**Review by vasinov — APPROVED** [review]

**Pull request merged**